File: leetcode/d/Word Break.java
/**
 * Given a string s and a dictionary of words dict, determine if s can be segmented into a space-separated sequence of one or more dictionary words.
 * <p>
 * For example, given
 * s = "leetcode",
 * dict = ["leet", "code"].
 * <p>
 * Return true because "leetcode" can be segmented as "leet code".
 */

public class Solution {

  // bottom up
  public boolean wordBreak(String s, Set<String> wordDict) {
    boolean[] canReach = new boolean[s.length()];
    for (int i = 0; i < canReach.length; i++) {
      if (i != 0 && !canReach[i - 1]) continue;
      for (int j = i + 1; j <= canReach.length; j++) {
        if (wordDict.contains(s.substring(i, j))) {
          canReach[j - 1] = true;
          if (j == canReach.length) return true;
        }
      }
    }
    return false;
  }

  // top down
  Map<String, Boolean> memo = new HashMap();

  public boolean wordBreak(String s, Set<String> wordDict) {
    if (s.length() == 0) return true;
    if (memo.containsKey(s)) return memo.get(s);
    // DP: string(0, n) = word(0,i) + string(i, n), for all possible word(0,i)
    // O(n^2), n number of sub-problems, each sub-problem is O(n), but O(2^n) if don't cache sub-problem
    for (int i = 1; i <= s.length(); i++) {
      if (wordDict.contains(s.substring(0, i))) {
        if (wordBreak(s.substring(i), wordDict)) {
          memo.put(s, true);
          return true;
        }
      }
    }
    memo.put(s, false);
    return false;
  }
}

File: lc_plusplus/Word Break.cc
// Given a non-empty string s and a dictionary wordDict containing a list of non-empty words, determine if s can be segmented into a space-separated sequence of one or more dictionary words. You may assume the dictionary does not contain duplicate words.
//
// For example, given
// s = "leetcode",
// dict = ["leet", "code"].
//
// Return true because "leetcode" can be segmented as "leet code".


class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> dict{wordDict.begin(), wordDict.end()};
        vector<bool> dp(s.size());
        for (size_t i = 0; i < dp.size(); ++i) {
            for (size_t len = 1; i + len <= dp.size(); ++len) {
                if ((i == 0 || dp[i - 1]) && dict.count(s.substr(i, len)) == 1) dp[i + len - 1] = true;
            }
        }
        return dp[s.size() - 1];
    }

    bool wordBreak(string s, vector<string>& wordDict) {
        if (s.empty()) return false;
        unordered_set<string> d{wordDict.begin(), wordDict.end()};
        vector<bool> dp(s.size());
        for (size_t i = 0; i < s.size(); ++i) {
            for (size_t j = i; j < s.size(); ++j) {
                if (i == 0 || dp[i - 1]) {
                    dp[j] = dp[j] || (d.count(s.substr(i, j - i + 1)) == 1);
                }
            }
        }
        return dp[s.size() - 1];
    }

    bool wordBreak2(string s, vector<string>& wordDict) {
        unordered_set<string> dict{wordDict.begin(), wordDict.end()};
        vector<bool> dp(s.size() + 1);
        dp[0] = true;
        for (int len{1}; len <= s.size(); ++len) {
            for (int i{0}; i < len; ++i) {
                if (dp[i]) {
                    auto w = s.substr(i, len - i);
                    if (dict.count(w) == 1) {
                        dp[len] = true;
                        break;
                    }
                }
            }
        }
        return dp[s.size()];
    }
};
